deepchem/feat/smiles_tokenizer.py

```diff
@@ -0,0 +1,292 @@
+            [(ids, tok) for tok, ids in self.vocab.items()]
+        )
+        self.basic_tokenizer = BasicSmilesTokenizer()
+        self.init_kwargs["max_len"] = self.max_len
+
+    @property
+    def vocab_size(self):
+        return len(self.vocab)
+
+    @property
+    def vocab_list(self):
+        return list(self.vocab.keys())
+
+    def _tokenize(self, text):
+        """
+        Tokenize a string into a list of tokens.
+
+        Parameters
+        ----------
+        text: str
+        """
+
+        split_tokens = [token for token in self.basic_tokenizer.tokenize(text)]
+        return split_tokens
+
+    def _convert_token_to_id(self, token):
+        """
+        Converts a token (str/unicode) in an id using the vocab.
+
+        Parameters
+        ----------
+        token: str
+        """
+
+        return self.vocab.get(token, self.vocab.get(self.unk_token))
+
+    def _convert_id_to_token(self, index):
+        """
+        Converts an index (integer) in a token (string/unicode) using the vocab.
+
+        Parameters
+        ----------
+        index: int
+        """
+
+        return self.ids_to_tokens.get(index, self.unk_token)
+
+    def convert_tokens_to_string(self, tokens):
+        """ Converts a sequence of tokens (string) in a single string.
+
+        Parameters
+        ----------
+        tokens: List[str]
+            List of tokens for a given string sequence.
+        """
+
+        out_string = " ".join(tokens).replace(" ##", "").strip()
+        return out_string
+
+    def add_special_tokens_ids_single_sequence(self, token_ids):
+        """
+        Adds special tokens to the a sequence for sequence classification tasks.
+        A BERT sequence has the following format: [CLS] X [SEP]
+
+        Parameters
+        ----------
+
+        token_ids: list[int]
+            list of tokenized input ids. Can be obtained using the encode or encode_plus methods.
+        """
+
+        return [self.cls_token_id] + token_ids + [self.sep_token_id]
+
+    def add_special_tokens_single_sequence(self, tokens):
+        """
+        Adds special tokens to the a sequence for sequence classification tasks.
+        A BERT sequence has the following format: [CLS] X [SEP]
+
+        Parameters
+        ----------
+        tokens: List[str]
+            List of tokens for a given string sequence.
+
+        """
+        return [self.cls_token] + tokens + [self.sep_token]
+
+    def add_special_tokens_sequence_pair(self, token_0, token_1):
+        """
+        Adds special tokens to a sequence pair for sequence classification tasks.
+        A BERT sequence pair has the following format: [CLS] A [SEP] B [SEP]
+
+        Parameters
+        ----------
+        token_0: str
+        token_1: str
+        """
+
+        sep = [self.sep_token]
+        cls = [self.cls_token]
+        return cls + token_0 + sep + token_1 + sep
+
+    def add_special_tokens_ids_sequence_pair(self, token_ids_0, token_ids_1):
```

Comment:
Can you add type annotations, return signatures, and a "Returns" section to the docstring here?

Reply:
done